File: src/BaseHmiTypes/DataLogging/HmiDataLog.cs
namespace BaseHmiTypes.DataLogging;

public sealed class HmiDataLog : IHmiObject
{
    public string? Name { get; set; }

    public string? Comment { get; set; }

    public string? Version { get; set; }

    public string? StorageLocation { get; set; }

    public string? ConnectionName { get; set; }

    public HmiDataLogRetentionPolicy RetentionPolicy { get; } = new();

    public IList<HmiDataLogGroup> Groups { get; } = new List<HmiDataLogGroup>();

    public IList<HmiDataLogTag> Tags { get; } = new List<HmiDataLogTag>();
}

public sealed class HmiDataLogRetentionPolicy : IHmiObject
{
    public bool Enabled { get; set; }

    public double? Duration { get; set; }

    public string? Unit { get; set; }
}

public sealed class HmiDataLogGroup : IHmiObject
{
    public string? Id { get; set; }

    public string? Name { get; set; }

    public HmiDataLogTrigger Trigger { get; } = new();
}

public sealed class HmiDataLogTag : IHmiObject
{
    public string? Name { get; set; }

    public string? Tag { get; set; }

    public string? DataType { get; set; }

    public string? GroupId { get; set; }

    public string? GroupName { get; set; }

    public HmiDataLogTrigger Trigger { get; } = new();
}

public sealed class HmiDataLogTrigger : IHmiObject
{
    public HmiDataLogTriggerType Type { get; set; }

    public double? Interval { get; set; }

    public string? IntervalUnit { get; set; }

    public double? MaximumUpdateRate { get; set; }

    public string? MaximumUpdateRateUnit { get; set; }

    public HmiDataLogDeadbandMode DeadbandMode { get; set; }

    public double? DeadbandValue { get; set; }

    public double? Heartbeat { get; set; }

    public string? HeartbeatUnit { get; set; }
}

public enum HmiDataLogTriggerType
{
    Unknown,
    Periodic,
    OnChange,
    OnDemand
}

public enum HmiDataLogDeadbandMode
{
    Unknown,
    None,
    Percentage,
    Absolute
}

File: typescript/base-hmi-types/src/data-logging/HmiDataLog.ts
export class HmiDataLog {
  name?: string;
  comment?: string;
  version?: string;
  storageLocation?: string;
  connectionName?: string;
  readonly retentionPolicy = new HmiDataLogRetentionPolicy();
  readonly groups: HmiDataLogGroup[] = [];
  readonly tags: HmiDataLogTag[] = [];
}

export class HmiDataLogRetentionPolicy {
  enabled = false;
  duration?: number;
  unit?: string;
}

export class HmiDataLogGroup {
  id?: string;
  name?: string;
  readonly trigger = new HmiDataLogTrigger();
}

export class HmiDataLogTag {
  name?: string;
  tag?: string;
  dataType?: string;
  groupId?: string;
  groupName?: string;
  readonly trigger = new HmiDataLogTrigger();
}

export class HmiDataLogTrigger {
  type = HmiDataLogTriggerType.Unknown;
  interval?: number;
  intervalUnit?: string;
  maximumUpdateRate?: number;
  maximumUpdateRateUnit?: string;
  deadbandMode = HmiDataLogDeadbandMode.Unknown;
  deadbandValue?: number;
  heartbeat?: number;
  heartbeatUnit?: string;
}

export enum HmiDataLogTriggerType {
  Unknown = "Unknown",
  Periodic = "Periodic",
  OnChange = "OnChange",
  OnDemand = "OnDemand",
}

export enum HmiDataLogDeadbandMode {
  Unknown = "Unknown",
  None = "None",
  Percentage = "Percentage",
  Absolute = "Absolute",
}
